Modbus Fix message_wait_milliseconds is no longer applied

peetersch opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add self._lock again to secure the a waiting time = message_wait_milliseconds between 2 requests is applied again.

Undo the following pull-request:
https://github.com/home-assistant/core/pull/151997



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #153707
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [X] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

peetersch committed:
Update modbus.py

See
- https://github.com/home-assistant/core/issues/153707

erwindouna commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -413,9 +414,11 @@ async def async_pb_call(
-            await asyncio.sleep(self._msg_wait)
-        return result
+        async with self._lock:
+            if not self._client:
```
Shouldn't this be just outside/before the lock? It's a fast path method and acquiring the lock an then to find out the client isn't initialized, is a waste of resources if you expect a lot of traffic. :)

peetersch replied:
I agree with you that it's better to put the check if the client has already been initialized outside the lock.
I updated the code and retested everything again. 
It all works fine.
Thx for the feedback.

erwindouna replied:
Would it still be justified if you do double-checking? One for a fast path, one for the hot path. I personally don't know the numbers of traffic through this code, so merely asking. :)

peetersch committed:
Update modbus.py

Improvement following a review of erwindouna.
It's better to put the check if the client has already been initialized outside the lock.
I retested after making this change and everything is still working fine.

joostlek reviewed (approved):
I guess it makes sense as the lock was said to move to the library, while the sleeping happens here, so the lock doesn't catch that anymore

joostlek reviewed (changes requested):
Ruff is failing, can you take a look?

peetersch committed:
Merge branch 'dev' into peetersch-patch-1

peetersch reviewed (commented):
Look ok

peetersch committed:
Update modbus.py

removed trailing spaces after self._lock = asyncio.Lock()

peetersch committed:
Update modbus.py

Remove trailing spaces

peetersch committed:
Merge branch 'dev' into peetersch-patch-1

peetersch commented:
OK. Indeed there were trailing spaces...
I removed them, so now the Ruff should be OK.
<img width="2313" height="962" alt="image" src="https://github.com/user-attachments/assets/558d607b-00bd-41c9-b51d-b9e4f5709a7f" />

erwindouna commented:
Don't forget to run `pre-commit run` before you submit. That will bring sanitized and standardized codes according to the standards of HA. :)
